src/adapter/src/catalog/open.rs

```diff
@@ -1717,6 +1717,16 @@ impl Catalog {
             migrated_builtins,
         })
     }
+
+    /// Politely releases all external resources that can only be released in an async context.
+    pub async fn expire(self) {
+        // If no one else holds a reference to storage, then clean up the storage resources.
+        // Otherwise, hopefully the other reference cleans up the resources when it's dropped.
+        if let Some(storage) = Arc::into_inner(self.storage) {
+            let storage = storage.into_inner();
+            storage.expire().await;
+        }
+    }
```

Comment:
do you think it's worth it to add a `Drop` impl for `Catalog` that emits a `tracing::error!` if we're ever dropped without first having called `fn expire(...)`?

Reply:
I can add a `warn`, but I think in the common case we will not call `expire` since there's no clean shutdown of `environmentd`.

Reply:
Hmm I seem to be getting some issue when trying to add a `Drop` impl,

```
error[E0509]: cannot move out of type `PersistHandle`, which implements the `Drop` trait
   --> src/catalog/src/persist.rs:531:9
    |
531 |         self.read_handle.expire().await;
    |         ^^^^^^^^^^^^^^^^
    |         |
    |         cannot move out of here
    |         move occurs because `__self.read_handle` has type `ReadHandle<StateUpdateKind, (), mz_repr::Timestamp, i64>`, which does not implement the `Copy` trait

error[E0509]: cannot move out of type `PersistHandle`, which implements the `Drop` trait
   --> src/catalog/src/persist.rs:532:9
    |
532 |         self.write_handle.expire().await;
    |         ^^^^^^^^^^^^^^^^^
    |         |
    |         cannot move out of here
    |         move occurs because `__self.write_handle` has type `WriteHandle<StateUpdateKind, (), mz_repr::Timestamp, i64>`, which does not implement the `Copy` trait

For more information about this error, try `rustc --explain E0509`.
error: could not compile `mz-catalog` (lib) due to 2 previous errors

Process finished with exit code 101

```

I guess you can't move a field out of the struct in a method if it also implements `Drop`. It's probably not worth it since we expect `expire` not to be called the majority of the time.